src/deploy.ts

```diff
@@ -1,209 +1,378 @@
+  ): Promise<HypNative> {
+    const router = await this.deployContractFromFactory(
+      chain,
+      new HypNative__factory(),
+      'HypNative',
+      [],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(config.mailbox, config.interchainGasPaymaster),
+    );
+    return router;
+  }
+
+  protected async deploySynthetic(
+    chain: ChainName,
+    config: HypERC20Config,
+  ): Promise<HypERC20> {
+    const router = await this.deployContractFromFactory(
+      chain,
+      new HypERC20__factory(),
+      'HypERC20',
+      [config.decimals],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(
+        config.mailbox,
+        config.interchainGasPaymaster,
+        config.totalSupply,
+        config.name,
+        config.symbol,
       ),
-      factories,
-      options,
     );
+    return router;
   }
 
   router(contracts: HyperlaneContracts<HypERC20Factories>) {
     return contracts.router;
   }
 
   async deployContracts(chain: ChainName, config: HypERC20Config) {
+    let router: HypERC20 | HypERC20Collateral | HypNative;
     if (isCollateralConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20Collateral__factory(),
-        'HypERC20Collateral',
-        [config.token],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
-    } else if (isSyntheticConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20__factory(),
-        'HypERC20',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(
-          config.mailbox,
-          config.interchainGasPaymaster,
-          config.totalSupply,
-          config.name,
-          config.symbol,
-        ),
-      );
-      return { router };
+      router = await this.deployCollateral(chain, config);
     } else if (isNativeConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypNative__factory(),
-        'HypNative',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
+      router = await this.deployNative(chain, config);
+    } else if (isSyntheticConfig(config)) {
+      router = await this.deploySynthetic(chain, config);
+    } else {
+      throw new Error('Invalid ERC20 token router config');
+    }
+    return { router };
+  }
+
+  async buildTokenMetadata(
+    configMap: ChainMap<TokenConfig>,
+  ): Promise<ChainMap<ERC20Metadata>> {
+    let tokenMetadata: ERC20Metadata | undefined;
+
+    for (const [chain, config] of Object.entries(configMap)) {
+      if (isCollateralConfig(config)) {
```

Comment:
Should we break after these conditions? Alternatively, combine some of this with `fetchMetadata` (pass the multiprovider) and you can early return

Reply:
I believe it is necessary to build the token metadata for the synthetics? Though I agree for readability it would be nice to separate that in different loops and make it explicit